### ISSUE TITLE
No error for overlapping overloads that vary in parameters with default values

### ISSUE THREAD

**Issue by brianschubert** [python/mypy#19803]
**Bug Report**

Mypy doesn't complain that the following overloads overlap with incompatible return types, but probably should:
```python
from typing import  overload

@overload
def foo(*, a: str = ...) -> str: ...
@overload
def foo(*, a: int | str = ...) -> int: ...
```
For comparison, pyright emits a diagnostic here ([playground](https://pyright-play.net/?pythonVersion=3.13&strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpRgBuApiADZgCGAJgFB0ACxZltdNJwUwYYAFACoANFCoAuKAGcYIKAF4oAOhUBKKAFoAfNNmSVSpiwrV6nbrwEixk1PgA%2BuuYoPrtmFDH0q6QA)). Mypy does emit an error if a default value for `a` is omitted in either overload.

**Comment by sterliakov** [comment]
It was intentionally changed a while ago. I also consider this a regression, but failed to convince @ilevkivskyi that this behavior should be reverted; mypy 1.10 and older reports incompatible overloads in your snippet as expected. See my report in #18581 and issues&PRs linked there.

Ough, sorry, I missed another aspect of this report: it's not only the default value that makes the signatures overlap, this false negative is worse than #18581

**Comment by ilevkivskyi** [comment]
This is actually not the same. Here the types of the argument types are overlapping, so this is actually a bug.
